File: clients/go/acteon/worker_test.go
package acteon

// Go SDK polling Worker tests.
//
// These tests drive the Worker against a stateful httptest mock of
// the `/v1/queues` surface. The contract under test: poll → handle
// → complete on success; handler errors fail with retryable=true;
// NonRetryable-wrapped errors fail with retryable=false; slow
// handlers get automatic heartbeats; Run honors ctx cancellation.

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"testing"
	"time"
)

// queueMockState records every worker-side report the mock queue
// server receives. All access goes through the mutex — the Worker
// hits the server from multiple goroutines.
type queueMockState struct {
	mu         sync.Mutex
	pollCount  int
	pollBodies []map[string]any
	heartbeats []map[string]any
	completes  []map[string]any
	fails      []map[string]any
}

func (s *queueMockState) snapshot() (heartbeats, completes, fails []map[string]any) {
	s.mu.Lock()
	defer s.mu.Unlock()
	return append([]map[string]any{}, s.heartbeats...),
		append([]map[string]any{}, s.completes...),
		append([]map[string]any{}, s.fails...)
}

// newQueueMockServer returns a mock queue server that serves `tasks`
// on the first poll and an empty batch on every poll after, and
// records heartbeat / complete / fail bodies. Caller must
// `defer teardown()`.
func newQueueMockServer(t *testing.T, tasks []map[string]any) (string, *queueMockState, func()) {
	t.Helper()
	state := &queueMockState{}
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var body map[string]any
		_ = json.NewDecoder(r.Body).Decode(&body)

		state.mu.Lock()
		defer state.mu.Unlock()
		w.Header().Set("Content-Type", "application/json")

		switch {
		case strings.HasSuffix(r.URL.Path, "/poll"):
			state.pollCount++
			state.pollBodies = append(state.pollBodies, body)
			batch := []map[string]any{}
			if state.pollCount == 1 {
				batch = tasks
			}
			_ = json.NewEncoder(w).Encode(map[string]any{"tasks": batch})
		case strings.HasSuffix(r.URL.Path, "/heartbeat"):
			state.heartbeats = append(state.heartbeats, body)
			_ = json.NewEncoder(w).Encode(taskWire(map[string]any{"status": "leased"}))
		case strings.HasSuffix(r.URL.Path, "/complete"):
			state.completes = append(state.completes, body)
			_ = json.NewEncoder(w).Encode(taskWire(map[string]any{"status": "completed"}))
		case strings.HasSuffix(r.URL.Path, "/fail"):
			state.fails = append(state.fails, body)
			_ = json.NewEncoder(w).Encode(taskWire(map[string]any{"status": "failed"}))
		default:
			w.WriteHeader(http.StatusNotFound)
			_ = json.NewEncoder(w).Encode(map[string]any{"error": "not found"})
		}
	}))
	return srv.URL, state, srv.Close
}

// leasedTaskWire builds a leased wire-form task for the mock poll
// response.
func leasedTaskWire(taskID, actionType, leaseToken string, payload map[string]any) map[string]any {
	return taskWire(map[string]any{
		"task_id":          taskID,
		"action_type":      actionType,
		"payload":          payload,
		"status":           "leased",
		"attempt":          1,
		"lease_token":      leaseToken,
		"lease_expires_at": time.Now().UTC().Add(time.Minute).Format(time.RFC3339),
	})
}

func newTestWorker(serverURL string) *Worker {
	c := NewClient(serverURL, WithAPIKey("k"))
	return NewWorker(c, WorkerConfig{
		Namespace: "ns",
		Tenant:    "tnt",
		Queue:     "q",
		WorkerID:  "w-1",
	})
}

func TestWorkerConfigDefaults(t *testing.T) {
	w := NewWorker(NewClient("http://localhost"), WorkerConfig{
		Namespace: "ns",
		Tenant:    "tnt",
		Queue:     "q",
	})
	if w.cfg.PollInterval != time.Second {
		t.Errorf("PollInterval default: got %v", w.cfg.PollInterval)
	}
	if w.cfg.LeaseSeconds != 60 {
		t.Errorf("LeaseSeconds default: got %d", w.cfg.LeaseSeconds)
	}
	if w.cfg.MaxConcurrent != 1 {
		t.Errorf("MaxConcurrent default: got %d", w.cfg.MaxConcurrent)
	}
}

func TestWorkerRunOnceRequiresScope(t *testing.T) {
	w := NewWorker(NewClient("http://localhost"), WorkerConfig{})
	if _, err := w.RunOnce(context.Background()); err == nil {
		t.Fatalf("expected error for missing namespace/tenant/queue")
	}
}

func TestWorkerHappyPathCompletesWithHandlerResult(t *testing.T) {
	url, state, teardown := newQueueMockServer(t, []map[string]any{
		leasedTaskWire("t-1", "send_email", "lease-1", map[string]any{"to": "a@b.c"}),
	})
	defer teardown()

	w := newTestWorker(url)
	var gotPayload map[string]any
	w.Register("send_email", func(_ context.Context, payload json.RawMessage) (any, error) {
		if err := json.Unmarshal(payload, &gotPayload); err != nil {
			return nil, err
		}
		return map[string]any{"sent": true}, nil
	})

	n, err := w.RunOnce(context.Background())
	if err != nil {
		t.Fatalf("run once: %v", err)
	}
	if n != 1 {
		t.Errorf("tasks handled: got %d", n)
	}
	if gotPayload["to"] != "a@b.c" {
		t.Errorf("handler payload: got %v", gotPayload)
	}

	_, completes, fails := state.snapshot()
	if len(fails) != 0 {
		t.Errorf("unexpected fail reports: %v", fails)
	}
	if len(completes) != 1 {
		t.Fatalf("complete reports: got %d", len(completes))
	}
	complete := completes[0]
	if complete["lease_token"] != "lease-1" {
		t.Errorf("complete lease_token: got %v", complete["lease_token"])
	}
	result, ok := complete["result"].(map[string]any)
	if !ok || result["sent"] != true {
		t.Errorf("complete result: got %v", complete["result"])
	}

	// Poll body must carry the worker's scope + identity.
	state.mu.Lock()
	poll := state.pollBodies[0]
	state.mu.Unlock()
	if poll["namespace"] != "ns" || poll["tenant"] != "tnt" || poll["worker_id"] != "w-1" {
		t.Errorf("poll body: got %v", poll)
	}
}

func TestWorkerHandlerErrorFailsRetryable(t *testing.T) {
	url, state, teardown := newQueueMockServer(t, []map[string]any{
		leasedTaskWire("t-1", "send_email", "lease-1", map[string]any{}),
	})
	defer teardown()

	w := newTestWorker(url)
	w.Register("send_email", func(context.Context, json.RawMessage) (any, error) {
		return nil, errors.New("smtp timeout")
	})

	if _, err := w.RunOnce(context.Background()); err != nil {
		t.Fatalf("run once: %v", err)
	}

	_, completes, fails := state.snapshot()
	if len(completes) != 0 {
		t.Errorf("unexpected complete reports: %v", completes)
	}
	if len(fails) != 1 {
		t.Fatalf("fail reports: got %d", len(fails))
	}
	fail := fails[0]
	if fail["error"] != "smtp timeout" || fail["lease_token"] != "lease-1" {
		t.Errorf("fail body: got %v", fail)
	}
	if fail["retryable"] != true {
		t.Errorf("plain handler errors must fail retryable: got %v", fail["retryable"])
	}
}

func TestWorkerNonRetryableErrorFailsTerminal(t *testing.T) {
	url, state, teardown := newQueueMockServer(t, []map[string]any{
		leasedTaskWire("t-1", "send_email", "lease-1", map[string]any{}),
	})
	defer teardown()

	w := newTestWorker(url)
	w.Register("send_email", func(context.Context, json.RawMessage) (any, error) {
		// Wrapping with %w along the way must not hide the marker.
		return nil, fmt.Errorf("handler: %w", NonRetryable(errors.New("bad address")))
	})

	if _, err := w.RunOnce(context.Background()); err != nil {
		t.Fatalf("run once: %v", err)
	}

	_, _, fails := state.snapshot()
	if len(fails) != 1 {
		t.Fatalf("fail reports: got %d", len(fails))
	}
	if fails[0]["retryable"] != false {
		t.Errorf("NonRetryable errors must fail terminal: got %v", fails[0]["retryable"])
	}
	if !strings.Contains(fails[0]["error"].(string), "bad address") {
		t.Errorf("fail error: got %v", fails[0]["error"])
	}
}

func TestWorkerUnregisteredActionTypeFailsRetryable(t *testing.T) {
	url, state, teardown := newQueueMockServer(t, []map[string]any{
		leasedTaskWire("t-1", "unknown_type", "lease-1", map[string]any{}),
	})
	defer teardown()

	w := newTestWorker(url)
	w.Register("send_email", func(context.Context, json.RawMessage) (any, error) {
		return nil, nil
	})

	if _, err := w.RunOnce(context.Background()); err != nil {
		t.Fatalf("run once: %v", err)
	}

	_, _, fails := state.snapshot()
	if len(fails) != 1 {
		t.Fatalf("fail reports: got %d", len(fails))
	}
	if fails[0]["retryable"] != true {
		t.Errorf("missing-handler failures must stay retryable: got %v", fails[0]["retryable"])
	}
	if !strings.Contains(fails[0]["error"].(string), "unknown_type") {
		t.Errorf("fail error must name the action type: got %v", fails[0]["error"])
	}
}

func TestWorkerHeartbeatsSlowHandler(t *testing.T) {
	url, state, teardown := newQueueMockServer(t, []map[string]any{
		leasedTaskWire("t-1", "slow", "lease-1", map[string]any{}),
	})
	defer teardown()

	w := newTestWorker(url)
	// Shrink the heartbeat cadence so the test stays fast; the
	// production cadence is LeaseSeconds/2.
	w.heartbeatInterval = 20 * time.Millisecond
	w.Register("slow", func(ctx context.Context, _ json.RawMessage) (any, error) {
		select {
		case <-time.After(150 * time.Millisecond):
		case <-ctx.Done():
			return nil, ctx.Err()
		}
		return map[string]any{"ok": true}, nil
	})

	if _, err := w.RunOnce(context.Background()); err != nil {
		t.Fatalf("run once: %v", err)
	}

	heartbeats, completes, fails := state.snapshot()
	if len(fails) != 0 || len(completes) != 1 {
		t.Fatalf("reports: %d completes, %d fails", len(completes), len(fails))
	}
	if len(heartbeats) == 0 {
		t.Fatalf("expected at least one heartbeat for a slow handler")
	}
	hb := heartbeats[0]
	if hb["lease_token"] != "lease-1" {
		t.Errorf("heartbeat lease_token: got %v", hb["lease_token"])
	}
	// extend_seconds must re-request the configured lease (default 60).
	if hb["extend_seconds"] != float64(60) {
		t.Errorf("heartbeat extend_seconds: got %v", hb["extend_seconds"])
	}
}

func TestWorkerFastHandlerSendsNoHeartbeat(t *testing.T) {
	url, state, teardown := newQueueMockServer(t, []map[string]any{
		leasedTaskWire("t-1", "fast", "lease-1", map[string]any{}),
	})
	defer teardown()

	w := newTestWorker(url)
	w.Register("fast", func(context.Context, json.RawMessage) (any, error) {
		return nil, nil
	})

	if _, err := w.RunOnce(context.Background()); err != nil {
		t.Fatalf("run once: %v", err)
	}

	heartbeats, _, _ := state.snapshot()
	// LeaseSeconds defaults to 60 → first heartbeat at 30s; a fast
	// handler must finish well before the ticker ever fires.
	if len(heartbeats) != 0 {
		t.Errorf("unexpected heartbeats: %v", heartbeats)
	}
}

func TestWorkerRunHonorsContextCancellation(t *testing.T) {
	url, state, teardown := newQueueMockServer(t, []map[string]any{
		leasedTaskWire("t-1", "send_email", "lease-1", map[string]any{}),
	})
	defer teardown()

	w := newTestWorker(url)
	w.cfg.PollInterval = 10 * time.Millisecond
	w.Register("send_email", func(context.Context, json.RawMessage) (any, error) {
		return map[string]any{"sent": true}, nil
	})

	ctx, cancel := context.WithCancel(context.Background())
	done := make(chan error, 1)
	go func() { done <- w.Run(ctx) }()

	// Let the loop drain the first batch and idle-poll a few times.
	time.Sleep(100 * time.Millisecond)
	cancel()

	select {
	case err := <-done:
		if !errors.Is(err, context.Canceled) {
			t.Errorf("Run must return ctx.Err(): got %v", err)
		}
	case <-time.After(2 * time.Second):
		t.Fatalf("Run did not return after cancellation")
	}

	_, completes, _ := state.snapshot()
	if len(completes) != 1 {
		t.Errorf("complete reports: got %d", len(completes))
	}
	state.mu.Lock()
	polls := state.pollCount
	state.mu.Unlock()
	if polls < 2 {
		t.Errorf("Run must keep polling after an empty batch: got %d polls", polls)
	}
}

func TestNonRetryableNilStaysNil(t *testing.T) {
	if NonRetryable(nil) != nil {
		t.Errorf("NonRetryable(nil) must be nil")
	}
	if IsNonRetryable(errors.New("plain")) {
		t.Errorf("plain errors must not read as non-retryable")
	}
	if !IsNonRetryable(NonRetryable(errors.New("x"))) {
		t.Errorf("wrapped errors must read as non-retryable")
	}
}

File: clients/java/src/test/java/com/acteon/client/WorkerTest.java
package com.acteon.client;

import com.acteon.client.exceptions.NonRetryableException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Java SDK polling Worker tests.
 *
 * <p>These tests drive the {@link Worker} against a stateful
 * in-process HttpServer mock of the {@code /v1/queues} surface,
 * mirroring the Go/Python/Node SDK worker tests. The contract under
 * test: poll → handle → complete on success; handler exceptions fail
 * with retryable=true; {@link NonRetryableException} fails with
 * retryable=false; unregistered action types fail retryable; slow
 * handlers get automatic heartbeats; run() exits cleanly on stop().
 */
class WorkerTest {
    private static final ObjectMapper MAPPER = JsonMapper.build();

    /**
     * Stateful mock queue server: serves {@code firstBatch} on the
     * first poll and an empty batch on every poll after, and records
     * every poll / heartbeat / complete / fail body it receives.
     */
    static final class QueueMock implements AutoCloseable {
        final HttpServer server;
        final AtomicInteger pollCount = new AtomicInteger(0);
        final List<Map<String, Object>> pollBodies = Collections.synchronizedList(new ArrayList<>());
        final List<Map<String, Object>> heartbeats = Collections.synchronizedList(new ArrayList<>());
        final List<Map<String, Object>> completes = Collections.synchronizedList(new ArrayList<>());
        final List<Map<String, Object>> fails = Collections.synchronizedList(new ArrayList<>());
        private final List<Map<String, Object>> firstBatch;

        QueueMock(List<Map<String, Object>> firstBatch) throws IOException {
            this.firstBatch = firstBatch;
            this.server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
            this.server.createContext("/", this::handle);
            // Heartbeats overlap with the handler's settle calls; a
            // multi-threaded executor keeps them from serializing.
            this.server.setExecutor(Executors.newCachedThreadPool());
            this.server.start();
        }

        String url() {
            return "http://127.0.0.1:" + server.getAddress().getPort();
        }

        private void handle(HttpExchange exchange) throws IOException {
            String path = exchange.getRequestURI().getPath();
            byte[] raw = exchange.getRequestBody().readAllBytes();
            Map<String, Object> body = raw.length == 0
                ? Map.of()
                : MAPPER.readValue(raw, new TypeReference<Map<String, Object>>() {});

            Object resp;
            int status = 200;
            if (path.endsWith("/poll")) {
                pollBodies.add(body);
                List<Map<String, Object>> batch =
                    pollCount.incrementAndGet() == 1 ? firstBatch : List.of();
                resp = Map.of("tasks", batch);
            } else if (path.endsWith("/heartbeat")) {
                heartbeats.add(body);
                resp = taskWire("hb", "x", "lease", "leased");
            } else if (path.endsWith("/complete")) {
                completes.add(body);
                resp = taskWire("done", "x", "lease", "completed");
            } else if (path.endsWith("/fail")) {
                fails.add(body);
                resp = taskWire("failed", "x", "lease", "failed");
            } else {
                status = 404;
                resp = Map.of("error", "not found");
            }
            byte[] respBody = MAPPER.writeValueAsBytes(resp);
            exchange.getResponseHeaders().set("Content-Type", "application/json");
            exchange.sendResponseHeaders(status, respBody.length);
            exchange.getResponseBody().write(respBody);
            exchange.close();
        }

        @Override
        public void close() {
            server.stop(0);
        }
    }

    /** Builds a wire-form task for mock responses. */
    static Map<String, Object> taskWire(String taskId, String actionType, String leaseToken, String status) {
        Map<String, Object> wire = new HashMap<>();
        wire.put("task_id", taskId);
        wire.put("queue", "q");
        wire.put("action_type", actionType);
        wire.put("payload", Map.of());
        wire.put("status", status);
        wire.put("attempt", 1);
        wire.put("max_attempts", 3);
        wire.put("lease_token", leaseToken);
        wire.put("lease_expires_at", Instant.now().plusSeconds(60).toString());
        wire.put("created_at", Instant.now().toString());
        wire.put("updated_at", Instant.now().toString());
        return wire;
    }

    /** Builds a leased wire-form task with a payload for mock poll responses. */
    static Map<String, Object> leasedTaskWire(String taskId, String actionType, String leaseToken,
                                              Map<String, Object> payload) {
        Map<String, Object> wire = taskWire(taskId, actionType, leaseToken, "leased");
        wire.put("payload", payload);
        return wire;
    }

    private static Worker.Builder testWorkerBuilder(ActeonClient client) {
        return Worker.builder(client, "ns", "tnt", "q").workerId("w-1");
    }

    // -------------------------------------------------------------------------
    // Construction
    // -------------------------------------------------------------------------

    @Test
    void builderRequiresScope() {
        try (ActeonClient client = new ActeonClient("http://localhost")) {
            assertThrows(IllegalArgumentException.class,
                () -> Worker.builder(client, "", "tnt", "q"));
            assertThrows(IllegalArgumentException.class,
                () -> Worker.builder(client, "ns", null, "q"));
            assertThrows(IllegalArgumentException.class,
                () -> Worker.builder(client, "ns", "tnt", ""));
        }
    }

    @Test
    void builderGeneratesWorkerIdWhenOmitted() {
        try (ActeonClient client = new ActeonClient("http://localhost");
             Worker worker = Worker.builder(client, "ns", "tnt", "q").build()) {
            assertTrue(worker.getWorkerId().startsWith("worker-"), worker.getWorkerId());
        }
    }

    // -------------------------------------------------------------------------
    // Dispatch + settlement
    // -------------------------------------------------------------------------

    @Test
    void happyPathCompletesWithHandlerResult() throws Exception {
        try (QueueMock mock = new QueueMock(List.of(
                leasedTaskWire("t-1", "send_email", "lease-1", Map.of("to", "a@b.c"))));
             ActeonClient client = new ActeonClient(mock.url());
             Worker worker = testWorkerBuilder(client).build()) {

            AtomicReference<String> gotTo = new AtomicReference<>();
            worker.register("send_email", payload -> {
                gotTo.set(payload.get("to").asText());
                return Map.of("sent", true);
            });

            int n = worker.runOnce();
            assertEquals(1, n);
            assertEquals("a@b.c", gotTo.get());

            assertEquals(0, mock.fails.size(), mock.fails.toString());
            assertEquals(1, mock.completes.size());
            Map<String, Object> complete = mock.completes.get(0);
            assertEquals("lease-1", complete.get("lease_token"));
            assertEquals("ns", complete.get("namespace"));
            assertEquals("tnt", complete.get("tenant"));
            @SuppressWarnings("unchecked")
            Map<String, Object> result = (Map<String, Object>) complete.get("result");
            assertEquals(Boolean.TRUE, result.get("sent"));

            // Poll body must carry the worker's scope, identity, and
            // the documented defaults (max_tasks=1, lease_seconds=60).
            Map<String, Object> poll = mock.pollBodies.get(0);
            assertEquals("ns", poll.get("namespace"));
            assertEquals("tnt", poll.get("tenant"));
            assertEquals("w-1", poll.get("worker_id"));
            assertEquals(1, poll.get("max_tasks"));
            assertEquals(60, poll.get("lease_seconds"));
        }
    }

    @Test
    void handlerExceptionFailsRetryable() throws Exception {
        try (QueueMock mock = new QueueMock(List.of(
                leasedTaskWire("t-1", "send_email", "lease-1", Map.of())));
             ActeonClient client = new ActeonClient(mock.url());
             Worker worker = testWorkerBuilder(client).build()) {

            worker.register("send_email", payload -> {
                throw new RuntimeException("smtp timeout");
            });

            assertEquals(1, worker.runOnce());

            assertEquals(0, mock.completes.size(), mock.completes.toString());
            assertEquals(1, mock.fails.size());
            Map<String, Object> fail = mock.fails.get(0);
            assertEquals("smtp timeout", fail.get("error"));
            assertEquals("lease-1", fail.get("lease_token"));
            assertEquals(Boolean.TRUE, fail.get("retryable"),
                "plain handler exceptions must fail retryable");
        }
    }

    @Test
    void nonRetryableExceptionFailsTerminal() throws Exception {
        try (QueueMock mock = new QueueMock(List.of(
                leasedTaskWire("t-1", "send_email", "lease-1", Map.of())));
             ActeonClient client = new ActeonClient(mock.url());
             Worker worker = testWorkerBuilder(client).build()) {

            worker.register("send_email", payload -> {
                throw new NonRetryableException("bad address");
            });

            assertEquals(1, worker.runOnce());

            assertEquals(1, mock.fails.size());
            Map<String, Object> fail = mock.fails.get(0);
            assertEquals(Boolean.FALSE, fail.get("retryable"),
                "NonRetryableException must fail terminal");
            assertTrue(((String) fail.get("error")).contains("bad address"), fail.toString());
        }
    }

    @Test
    void unregisteredActionTypeFailsRetryable() throws Exception {
        try (QueueMock mock = new QueueMock(List.of(
                leasedTaskWire("t-1", "unknown_type", "lease-1", Map.of())));
             ActeonClient client = new ActeonClient(mock.url());
             Worker worker = testWorkerBuilder(client).build()) {

            worker.register("send_email", payload -> null);

            assertEquals(1, worker.runOnce());

            assertEquals(1, mock.fails.size());
            Map<String, Object> fail = mock.fails.get(0);
            assertEquals(Boolean.TRUE, fail.get("retryable"),
                "missing-handler failures must stay retryable");
            assertTrue(((String) fail.get("error")).contains("unknown_type"),
                "fail error must name the action type: " + fail.get("error"));
        }
    }

    // -------------------------------------------------------------------------
    // Heartbeats
    // -------------------------------------------------------------------------

    @Test
    void heartbeatsFireForSlowHandler() throws Exception {
        try (QueueMock mock = new QueueMock(List.of(
                leasedTaskWire("t-1", "slow", "lease-1", Map.of())));
             ActeonClient client = new ActeonClient(mock.url());
             // Shrink the heartbeat cadence so the test stays fast;
             // the production cadence is leaseSeconds/2.
             Worker worker = testWorkerBuilder(client)
                 .heartbeatInterval(Duration.ofMillis(20))
                 .build()) {

            worker.register("slow", payload -> {
                Thread.sleep(150);
                return Map.of("ok", true);
            });

            assertEquals(1, worker.runOnce());

            assertEquals(0, mock.fails.size(), mock.fails.toString());
            assertEquals(1, mock.completes.size());
            assertFalse(mock.heartbeats.isEmpty(),
                "expected at least one heartbeat for a slow handler");
            Map<String, Object> hb = mock.heartbeats.get(0);
            assertEquals("lease-1", hb.get("lease_token"));
            // extend_seconds must re-request the configured lease (default 60).
            assertEquals(60, hb.get("extend_seconds"));
        }
    }

    @Test
    void fastHandlerSendsNoHeartbeat() throws Exception {
        try (QueueMock mock = new QueueMock(List.of(
                leasedTaskWire("t-1", "fast", "lease-1", Map.of())));
             ActeonClient client = new ActeonClient(mock.url());
             Worker worker = testWorkerBuilder(client).build()) {

            worker.register("fast", payload -> null);

            assertEquals(1, worker.runOnce());

            // leaseSeconds defaults to 60 → first heartbeat at 30s; a
            // fast handler finishes well before the schedule fires.
            assertEquals(0, mock.heartbeats.size(), mock.heartbeats.toString());
            assertEquals(1, mock.completes.size());
        }
    }

    // -------------------------------------------------------------------------
    // Run loop + stop
    // -------------------------------------------------------------------------

    @Test
    void runProcessesTasksAndStopsCleanly() throws Exception {
        try (QueueMock mock = new QueueMock(List.of(
                leasedTaskWire("t-1", "send_email", "lease-1", Map.of())));
             ActeonClient client = new ActeonClient(mock.url());
             Worker worker = testWorkerBuilder(client)
                 .pollInterval(Duration.ofMillis(10))
                 .build()) {

            worker.register("send_email", payload -> Map.of("sent", true));

            Thread runner = new Thread(worker::run, "worker-run-test");
            runner.start();

            // Let the loop drain the first batch and idle-poll a few times.
            Thread.sleep(150);
            worker.stop();
            runner.join(2_000);
            assertFalse(runner.isAlive(), "run() did not return after stop()");

            assertEquals(1, mock.completes.size());
            assertTrue(mock.pollCount.get() >= 2,
                "run() must keep polling after an empty batch: got " + mock.pollCount.get());
        }
    }

    @Test
    void runOncePollFailureSurfacesAsException() throws Exception {
        HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
        server.createContext("/", exchange -> {
            byte[] respBody = "{\"error\":\"backend unavailable\"}".getBytes(StandardCharsets.UTF_8);
            exchange.getResponseHeaders().set("Content-Type", "application/json");
            exchange.sendResponseHeaders(503, respBody.length);
            exchange.getResponseBody().write(respBody);
            exchange.close();
        });
        server.start();
        try {
            int port = server.getAddress().getPort();
            try (ActeonClient client = new ActeonClient("http://127.0.0.1:" + port);
                 Worker worker = Worker.builder(client, "ns", "tnt", "q").build()) {
                assertThrows(com.acteon.client.exceptions.ActeonException.class, worker::runOnce);
            }
        } finally {
            server.stop(0);
        }
    }
}
